File: cpp_version/src/client/loginWindow.cpp
#include "../include/loginWindow.h"
#include "../build/message.pb.h"

#include <iostream>
#include <stdio.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <errno.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <netinet/in.h>
#include <signal.h>
#include <arpa/inet.h>
#include <thread>


#define MAXSIZE 1024
#define IP_ADDR "127.0.0.1"
#define IP_PORT 8888



std::string LoginDataSerialize(const std::string& userid, const std::string& passwd){
	std::string logindata;
	message::LoginMsg loginmsg;
	loginmsg.set_userid(userid);
	loginmsg.set_userpwd(passwd);
	loginmsg.set_username("xieyibo0112");
	loginmsg.SerializeToString(&logindata);
	return logindata;
};


bool LoginWindow::LoginCheck(const std::string& userid, const std::string& passwd){
	bool flag;
	struct sockaddr_in st_clnsock;
	int nrecvSize = 0;

	if((i_sockfd = socket(AF_INET, SOCK_STREAM, 0) ) < 0)	//建立套接字
	{
		printf("socket Error: %s (errno: %d)\n", strerror(errno), errno);
		exit(0);
	}

	memset(&st_clnsock, 0, sizeof(st_clnsock));
	st_clnsock.sin_family = AF_INET;  //IPv4协议
	//IP地址转换(直接可以从物理字节序的点分十进制 转换成网络字节序)
	if(inet_pton(AF_INET, IP_ADDR, &st_clnsock.sin_addr) <= 0)
	{
		printf("inet_pton Error: %s (errno: %d)\n", strerror(errno), errno);
		exit(0);
	}
	st_clnsock.sin_port = htons(IP_PORT);	//端口转换(物理字节序到网络字节序)

	if(connect(i_sockfd, (struct sockaddr*)&st_clnsock, sizeof(st_clnsock)) < 0)	//主动向设置的IP和端口号的服务端发出连接
	{
		printf("connect Error: %s (errno: %d)\n", strerror(errno), errno);
		exit(0);
	}

	printf("======connect to server, sent data======\n");

	// 传输登录信息
	tf = new Transfer(i_sockfd); 
	tf->sendMessage(message::MSGTYPE::LoginMsgType, LoginDataSerialize(userid, passwd), LoginDataSerialize(userid, passwd).length());

	// 等待客户端回复
	message::Message msg = tf->readMessage();
	message::ResponseMsg resMsg;
	bool ret2 = resMsg.ParseFromString(msg.data());
	if (ret2 && resMsg.code() == "200"){
		// std::cout << "登录成功" << std::endl;
		flag = true;
	}else{
		flag = false;
	}
	// printf("======recv finished======\n");
	
	google::protobuf::ShutdownProtobufLibrary();
	return flag;
}

void LoginWindow::LoginShow(){
    std::cout << "请输入用户的id" << std::endl;
	std::cin >> this->userid;
	std::cout <<"请输入用户的密码"<< std::endl;
	std::cin >> this->passwd;
	bool ret = this->LoginCheck(this->userid, this->passwd);
	
	if (!ret) {
		std::cout <<"登录失败" << std::endl;
		close(i_sockfd);
	} else {
		std::cout <<"登录成功" << std::endl;
		lsw = new LoginSuccWindow();
		std::thread t1 = std::thread(&LoginSuccWindow::RecvMsgFromSrv, lsw, i_sockfd);
		lsw->ShowMenu();
		t1.join();	
	}
	
}

File: go_version/client/loginWindow/loginWindow.go
package loginWindow

import (
	"commsys/client/process"
	"commsys/common/message"
	"commsys/common/utils"
	"encoding/json"
	"errors"
	"fmt"
	"net"
)

type LoginWindow struct {
	userid string
	passwd string
	tf     *utils.Transfer
}

func (lw *LoginWindow) LoginCheck(userid string, passwd string) (err error) {
	// 开始制定协议。。。
	// fmt.Printf("userId = %s userpwd = %s\n", userid, passwd)
	// 1. 连接到服务器
	conn, err := net.Dial("tcp", "localhost:8889")
	if err != nil {
		fmt.Println("Dial err = ", err)
	}

	// 延时关闭
	defer conn.Close()

	// 2. 准备通过conn发送消息给服务
	var msg message.Message

	// 3. 创建一个LoginMsg结构体
	var loginMsg message.LoginMsg
	loginMsg.UserId = userid
	loginMsg.UserPwd = passwd

	// 4. 将loginMsg序列化
	data, err := json.Marshal(loginMsg)
	// fmt.Printf("%c\n", data)
	if err != nil {
		fmt.Println("json.Marshal err = ", err)
		return err
	}

	// 5. 把data赋给msg.Data字段，把长度赋给msg.Length字段
	msg.Data = string(data)
	msg.Length = len(msg.Data)
	msg.Type = message.LoginMsgType

	// 构建transfer实例
	lw.tf.Conn = conn
	// 6. 将msg发送到服务器
	lw.tf.WritePkg(msg)

	fmt.Println("客户端发送消息成功")

	// 处理服务器返回的消息

	ResMsg, err := lw.tf.ReadPkg()

	// 反序列化
	var loginRes message.ResponseMsg
	err = json.Unmarshal([]byte(ResMsg.Data), &loginRes)
	if err != nil {
		fmt.Println("json unmarshal err = ", err)
		return
	}

	if loginRes.Code == 200 {
		lsw := &process.LoginSuccWindow{
			Loop:      true,
			Tf:        lw.tf,
			SrcUserId: userid,
		}

		// 启动一个协程，保持和服务器端通讯，如果服务器有数据推送给客户端，可以接收并显示在客户端的终端
		go lsw.RecvMsgFromServer(conn)
		// 显示登录成功后的二级菜单
		lsw.ShowMenu()
		return
	} else {
		fmt.Println(loginRes.Error)
		return errors.New("账号密码错误")
	}
}

func (lw *LoginWindow) LoginShow() {
	fmt.Println("请输入用户的id")
	fmt.Scanf("%s\n", &lw.userid)
	fmt.Println("请输入用户的密码")
	fmt.Scanf("%s\n", &lw.passwd)
	lw.tf = &utils.Transfer{}
	err := lw.LoginCheck(lw.userid, lw.passwd)
	if err != nil {
		fmt.Println("登录失败")
	} else {
		fmt.Println("登录成功")
	}
}
